File: 图论算法/图上操作/最短路/Floyd/Floyd.cpp
#include <bits/stdc++.h>
using namespace std;
//Floyd算法用于求全源最短路,时间复杂度O(n^3),可以求任意两点间最短距离,可以求解负权图
//start,end,weight分别代表边的起点,终点及边权,n代表点的数量,返回值为nxn数组,d[i][j]表示i~j之间最短路
vector<vector<int>> Floyd(vector<int>start,vector<int>end,vector<int>weight,int n){
    const int INF = 0x3f3f3f3f;
    vector<vector<int>>d(n,vector<int>(n,INF));
    for (int i=0;i<n;i++) d[i][i]=0;
    for (int i=0;i<start.size();i++){
        int x=start[i];
        int y=end[i];
        int z=weight[i];
        d[x][y]=min(z,d[x][y]);
    }
    for (int k=0;k<n;k++)
        for (int i=0;i<n;i++){
            if (d[i][k]==INF) continue;
            for (int j=0;j<n;j++)
                d[i][j]=min(d[i][j],d[i][k]+d[k][j]);
        }
    return d;
}

//实例:https://leetcode.cn/problems/minimum-cost-to-convert-string-i/description/

File: 图论算法/图上操作/最短路/Floyd/Floyd.py
from typing import List
from bisect import bisect_left, bisect_right
from heapq import heappush, heappop, heapreplace
from itertools import permutations, accumulate
from math import inf, comb, sqrt, ceil, floor, log, log2, log10
from functools import cache
from math import gcd, isqrt
from collections import defaultdict, deque, Counter

# from sortedcontainers import SortedList
# from itertools import pairwise

fmax = lambda x, y: x if x > y else y
fmin = lambda x, y: x if x < y else y

#Floyd算法用于求全源最短路,时间复杂度O(n^3),可以求任意两点间最短距离,可以求解负权图
#start,end,weight分别代表边的起点,终点及边权,n代表点的数量,返回值为nxn数组,d[i][j]表示i~j之间最短路
def Floyd(start,end,weight,n):
    d = [[inf] * n for _ in range(n)]
    for i in range(n):
        d[i][i] = 0

    for x, y, z in zip(start,end,weight):
        d[x][y] = fmin(z, d[x][y])

    for k in range(n):
        for i in range(n):
            if d[i][k] == inf:
                continue
            for j in range(n):
                d[i][j] = fmin(d[i][j], d[i][k] + d[k][j])
    return d